x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/snapshotlifecycle/SnapshotLifecyclePolicyMetadata.java

```diff
@@ -150,4 +229,57 @@ public String toString() {
         // should not emit them in case it accidentally gets logged.
         return super.toString();
     }
+
+    public static class Builder {
+
+        private Builder() {}
+
+        private SnapshotLifecyclePolicy policy;
+        private Map<String, String> headers;
+        private long version = 1L;
+        private Long modifiedDate;
+        private Long lastSuccessDate;
+        private Long lastFailureDate;
+        private String lastFailureInfo;
+
+        public Builder setPolicy(SnapshotLifecyclePolicy policy) {
+            this.policy = policy;
+            return this;
+        }
+
+        public Builder setHeaders(Map<String, String> headers) {
+            this.headers = headers;
+            return this;
+        }
+
+        public Builder setVersion(long version) {
+            this.version = version;
+            return this;
+        }
+
+        public Builder setModifiedDate(long modifiedDate) {
+            this.modifiedDate = modifiedDate;
+            return this;
+        }
+
+        public Builder setLastSuccessDate(Long lastSuccessDate) {
+            this.lastSuccessDate = lastSuccessDate;
+            return this;
+        }
+
+        public Builder setLastFailureDate(Long lastFailureDate) {
+            this.lastFailureDate = lastFailureDate;
+            return this;
+        }
+
+        public Builder setLastFailureInfo(String lastFailureInfo) {
+            this.lastFailureInfo = lastFailureInfo;
+            return this;
+        }
+
+        public SnapshotLifecyclePolicyMetadata build() {
+            Objects.requireNonNull(modifiedDate, "modifiedDate must be set");
```

Comment:
`policy` is null-checked in the constructor - I can move that check here, I only split them because I originally did it in the constructor but a null `modifiedDate` would throw an NPE at the time it gets passed into the constructor. I'll do all the checks in one place.

Good point on `headers`, I'll do that too.